### DISCUSSION TITLE
About takeoff command on CF Bolt

### DISCUSSION THREAD

**Discussion by Williamwenda** [USC-ACTLab/crazyswarm#375] [Q&A] [answered]
Hello,

Recently, we built a customized drone with CF Bolt and tested it with crazyswarm code. Everything is working pretty good except the takeoff process. With the CF Bolt drone, the take-off is pretty aggressive, which is not the case for a crazyflie. A video is [here](https://drive.google.com/file/d/1ML40TUIw2Ho40JKES5kk6cOp8UFuRTvu/view?usp=sharing).

The takeoff in crazyswarm code is sending a high-level command to the drone and the onboard firmware is running the takeoff process (in _planner.c_, function  _plan_takeoff_or_landing_, which calls the _piecewise_plan_7th_order_no_jerk_ in _pptraj.c_). 

However, it doesn't explain the aggressive jump at the beginning of the trajectory. Do you have any ideas why the jump happens for CF Bolt-based drone at the beginning of take-off?

Thanks a lot!

Kind regards,

Wenda

**Comment by jpreiss** [comment]
Please make your video public.

The most likely problem is that the feed-forward control input is wrong, so the CF falls behind the planned trajectory in the beginning. The Mellinger feedback controller is very aggressive and causes the jump to compensate for the tracking error. Make sure you adjust the `mass` and `massThrust` constants for your hardware.

**Reply by Williamwenda** [reply]
Thanks for your quick reply. Here is the [video](https://drive.google.com/drive/folders/1qG3p1f7HJ5ebk628pIyv8hlFYa3_Tbz5). The link should work this time
I am using PID controller (using config.mk to set the PID controller) and set the right mass. But I didn't change the massThrust in the Mellinger controller. Is the takeoff process using Mellinger as default?

**Reply by jpreiss** [reply]
I've never seen behavior like this.  It almost seems like something weird is going on with your ESCs. The climb and hover after the initial disturbance look pretty good - I would expect to see more problems if there was something seriously wrong with the controller setup.

The takeoff uses the same controller as everything else.

How are you using config.mk to set the PID controller? Our standard process is to compile all the controllers and use the parameter system + launch file to select the one we want at runtime.

**Reply by jpreiss** [reply]
I also suspect that this issue has nothing to do with Crazyswarm. You might try asking the same question in the Bitcraze forums for the Bolt.

**Reply by Williamwenda** [reply]
_How are you using config.mk to set the PID controller? Our standard process is to compile all the controllers and use the parameter system + launch file to select the one we want at runtime._
Interesting! I didn't notice that param in the launch file and the default is mellinger. But I guess I only compiled PID controller by setting config.mk so that I was still using PID. 

_The climb and hover after the initial disturbance look pretty good_ 
All other parts are pretty smooth. I also tested to fly the CF Bolt drone for a simple circle trajectory and it works pretty well. 

I guess one potentially problem could be the initial z position of the drone is around 0.1. But the initial z position I set was 0. The z offset might cause a large control input. Yet, I haven't seen a similar performance with crazyflie.

Anyway, thanks for your reply. It's good to know the takeoff and landing are using the same controller as everything else. I will further investigate the takeoff problem.

**Reply by jpreiss** [reply]
The initial z position shouldn't matter. The takeoff is always planned from the current position.

**Comment by Williamwenda** [comment, marked as answer]
Hello,

An update for the CF Bolt takeoff issue. I found the problem about that is due to the onboard thrust parameter. 
By changing the [thrustBase](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/position_controller_pid.c#L135), the CF Bolt drone doesn't jump aggressively during the takeoff process.

The problem is not from the ground station planning side. 
I'm appreciated for your help, jpreiss.

**Reply by jpreiss** [reply]
Great, thanks for sharing your solution! By the way, this is the equivalent of the `massThrust` tuning i mentioned earlier, but for the PID controller instead of Mellinger.

**Reply by Williamwenda** [reply]
Yes, that's true. 
`massThrust` is for Mellinger controller.
`thurstBase` is for PID controller.
